rfc3195d.8
.\" Copyright 2005 Rainer Gerhards and Adiscon for the rsyslog modifications
.\" Distributed under the GNU General Public License.
.\"
.TH RFC3195D 8 "02 April 2008" "Version 3.14.0" "Linux System Administration"
.SH NAME
rfc3195d \- RFC 3195 listener
.SH SYNOPSIS
.B rfc3195d
.RB [ " \-d " ]
.RB [ " \-p"
.IB socket 
]
.RB [ " \-r"
.IB port 
]
.RB [ " \-v " ]
.LP
.SH DESCRIPTION
.B Rfc3195d
is a utility for receiving syslog messages via RFC 3195. Both
RAW and COOKED profiles are supported (but COOKED only without
relay-specific PATH elements).
rfc3195d accepts messages via RFC 3195 and forwards them to
the local domain socket specified in the -p option
(/dev/log3195 by default). There, the messages can be picked up
by the system syslogd. While rfc3195d can work with any syslogd,
we highly recommend using
.B rsyslogd,
because it has special handling
for the messages forwarded by rfc3195d. To enable message 
reception in
.B rsyslogd,
use the "-a :/dev/log3195" command line
option (the colon in front of the socket name tells
.B rsyslogd
that the messages contain hostnames - this is vital to get the
right sender name into your logs).

.B Rfc3195d
currently has very limited functionality. Most importantly,
it does not allow to limit the senders it receives messages from.
Documentation is also very sparse. The situation should improve over
time as the rsyslog project is continously being enhanced.
.LP
.SH OPTIONS
.TP
.BI "\-p " "socket"
The socket the received messages are to be sent to. If not specified,
/dev/log3195 is used.
.TP
.BI "\-r " "port"
The listen port to use. If not specified, IANA-assigned port 601 is used.
.TP
.B "\-d"
Turns on debug mode.  In it, rfc3195d spits out diagnostic information
to stdout.
.TP
.B "\-v"
Print version and exit.
.SH SIGNALS
.B Rfc3195d
reacts to a set of signals.
.TP
.B SIGTERM
.B Rfc3195d
terminates.
.TP
.B SIGUSR1
.B Rfc3195d
terminates.
.LP
.SH SEE ALSO
.BR rsyslog.conf (5),
.BR rsyslogd (8)
.LP
.SH MORE INFORMATION
Is available on the project home page at http://www.rsyslog.com
.LP
.SH COLLABORATORS
Rfc3195d uses liblogging (http://www.liblogging.org) for RFC 3195
protocol handling.
.PD 0
.TP
Rainer Gerhards <rgerhards@adiscon.com>


NEWS
---------------------------------------------------------------------------
Version 1.15.1 (RGer), 2007-07-10
- fixed a bug that caused a dynaFile selector to stall when there was
  an open error with one file 
- improved template processing for dynaFiles; templates are now only
  looked up during initialization - speeds up processing
- optimized memory layout in struct filed when compiled with MySQL
  support
- fixed a bug that caused compilation without SYSLOG_INET to fail
- re-enabled the "last message repeated n times" feature. This
  feature was not taken care of while rsyslogd evolved from sysklogd
  and it was more or less defunct. Now it is fully functional again.
- added system properties: $NOW, $YEAR, $MONTH, $DAY, $HOUR, $MINUTE
- fixed a bug in iovAsString() that caused a memory leak under stress
  conditions (most probably memory shortage). This was unlikely to
  ever happen, but it doesn't hurt doing it right
- cosmetic: defined type "uchar", change all unsigned chars to uchar
---------------------------------------------------------------------------
Version 1.15.0 (RGer), 2007-07-05
- added ability to dynamically generate file names based on templates
  and thus properties. This was a much-requested feature. It makes
  life easy when it e.g. comes to splitting files based on the sender
  address.
- added $umask and $FileCreateMode config file directives
- applied a patch from Bartosz Kuzma to compile cleanly under NetBSD
- checks for extra (unexpected) characters in system config file lines
  have been added
- added IPv6 documentation - was accidently missing from CVS
- begun to change char to unsigned char
---------------------------------------------------------------------------
Version 1.14.2 (RGer), 2007-07-03
** this release fixes all known nits with IPv6 **
- restored capability to do /etc/service lookup for "syslog"
  service when -r 0 was given
- documented IPv6 handling of syslog messages
- integrate patch from Bartosz Kuźma to make rsyslog compile under
  Solaris again (the patch replaced a strndup() call, which is not
  available under Solaris
- improved debug logging when waiting on select
- updated rsyslogd man page with new options (-46A)
---------------------------------------------------------------------------
Version 1.14.1 (RGer/Peter Vrabec), 2007-06-29
- added Peter Vrabec's patch for IPv6 TCP
- prefixed all messages send to stderr in rsyslogd with "rsyslogd: "
---------------------------------------------------------------------------
Version 1.14.0 (RGer/Peter Vrabec), 2007-06-28
- Peter Vrabec provided IPv6 for rsyslog, so we are now IPv6 enabled
  IPv6 Support is currently for UDP only, TCP is to come soon.
  AllowedSender configuration does not yet work for IPv6.
- fixed code in iovCreate() that broke C's strict aliasing rules 
- fixed some char/unsigned char differences that forced the compiler
  to spit out warning messages
- updated the Red Hat init script to fix a known issue (thanks to
  Peter Vrabec)
---------------------------------------------------------------------------
Version 1.13.5 (RGer), 2007-06-22
- made the TCP session limit configurable via command line switch
  now -t <port>,<max sessions>
- added man page for rklogd(8) (basically a copy from klogd, but now
  there is one...)
- fixed a bug that caused internal messages (e.g. rsyslogd startup) to
  appear without a tag.
- removed a minor memory leak that occurred when TAG processing requalified
  a HOSTNAME to be a TAG (and a TAG already was set).
- removed potential small memory leaks in MsgSet***() functions. There
  would be a leak if a property was re-set, something that happened
  extremely seldom.
---------------------------------------------------------------------------
Version 1.13.4 (RGer), 2007-06-18
- added a new property "PRI-text", which holds the PRI field in
  textual form (e.g. "syslog.info")
- added alias "syslogseverity" for "syslogpriority", which is a
  misleading property name that needs to stay for historical
  reasons (and backward-compatility)
- added doc on how to record PRI value in log file
- enhanced signal handling in klogd, including removal of an unsafe
  call to the logging system during signal handling
---------------------------------------------------------------------------
Version 1.13.3 (RGer), 2007-06-15
- create a version of syslog.c from scratch. This is now
  - highly optimized for rsyslog
  - removes an incompatible license problem as the original
    version had a BSD license with advertising clause
  - fixed in the regard that rklogd will continue to work when
    rsysogd has been restarted (the original version, as well
    as sysklogd, will remain silent then)
  - solved an issue with an extra NUL char at message end that the
    original version had
- applied some changes to klogd to care for the new interface
- fixed a bug in syslogd.c which prevented compiling under debian
---------------------------------------------------------------------------
Version 1.13.2 (RGer), 2007-06-13
- lib order in makefile patched to facilitate static linking - thanks
  to Bennett Todd for providing the patch
- Integrated a patch from Peter Vrabec (pvrabec@redheat.com):
  - added klogd under the name of rklogd (remove dependency on
    original sysklogd package
  - createDB.sql now in UTF
  - added additional config files for use on Red Hat
---------------------------------------------------------------------------
Version 1.13.1 (RGer), 2007-02-05
- changed the listen backlog limit to a more reasonable value based on
  the maximum number of TCP connections configurd (10% + 5) - thanks to Guy
  Standen for the hint (actually, the limit was 5 and that was a 
  left-over from early testing).
- fixed a bug in makefile which caused DB-support to be disabled when
  NETZIP support was enabled
- added the -e option to allow transmission of every message to remote
  hosts (effectively turns off duplicate message suppression)

rfc3195d.c
/** 
 * rfc3195d.c
 * This is an RFC 3195 listener. All data received is forwarded to
 * local UNIX domain socket, where it can be picked up by a
 * syslog daemon (like rsyslogd ;)).
 *
 * \author  Rainer Gerhards <rgerhards@adiscon.com>
 *
 * Copyright 2003-2005 Rainer Gerhards and Adiscon GmbH.
 *
 * This file is part of rsyslog.
 *
 * Rsyslog is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Rsyslog is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Rsyslog.  If not, see <http://www.gnu.org/licenses/>.
 *
 * A copy of the GPL can be found in the file "COPYING" in this distribution.
 */
#include "config.h"

#include <stdio.h>
#ifndef FEATURE_RFC3195
/* this is a trick: if RFC3195 is not to be supported, we just do an
 * error message.
 */
int main()
{
	fprintf(stderr, "error: not compiled with FEATURE_RFC3195 - terminating.\n");
	return(1);
}
#else
#include <unistd.h>
#include <signal.h>
#include <sys/socket.h>
#include <sys/errno.h>
#include "rsyslog.h"
#include "liblogging.h"
#include "srAPI.h"
#include "syslogmessage.h"

/* configurable params! */
static char* pPathLogname = "/dev/log3195";
static char *PidFile;
static int NoFork = 0;
static int Debug = 0;
static int listenPort = 601;

/* we use a global API object below, because this listener is
 * not very complex. As such, this hack should not harm anything.
 * rgerhards, 2005-10-12
 */
static srAPIObj* pAPI;

static int	LogFile = -1;		/* fd for log */
static int	connected;		/* have done connect */
static struct sockaddr SyslogAddr;	/* AF_UNIX address of local logger */

/* small usage info */
static int usage()
{
	/* The following usage line is what we intend to have - it
	 * is commented out as a reminder. The one below is what we
	 * currently actually do...
	 fprintf(stderr, "usage: rfc3195d [-dv] [-i pidfile] [-n] [-p path]\n");
	 */
	fprintf(stderr, "usage: rfc3195d [-dv] [-r port] [-p path]\n");
	exit(1);
}

/* CLOSELOG -- close the system log
 */
static void closelog(void)
{
	close(LogFile);
	LogFile = -1;
	connected = 0;
}

/* OPENLOG -- open system log
 */
static void openlog()
{
	if (LogFile == -1) {
		SyslogAddr.sa_family = AF_UNIX;
		strncpy(SyslogAddr.sa_data, pPathLogname,
		    sizeof(SyslogAddr.sa_data));
		LogFile = socket(AF_UNIX, SOCK_DGRAM, 0);
		if(LogFile < 0) {
			char errStr[1024];
			printf("error opening '%s': %s\n", 
			       pPathLogname, rs_strerror_r(errno, errStr, sizeof(errStr)));
		}
	}
	if (LogFile != -1 && !connected &&
	    connect(LogFile, &SyslogAddr, sizeof(SyslogAddr.sa_family)+
			strlen(SyslogAddr.sa_data)) != -1)
		connected = 1;
	else {
		char errStr[1024];
		printf("error connecting '%s': %s\n", 
		       pPathLogname, rs_strerror_r(errno, errStr, sizeof(errStr)));
	}
}


/* This method is called when a message has been fully received.
 * It passes the received message to the specified unix domain
 * socket. Please note that this callback is synchronous, thus
 * liblogging will be on hold until it returns. This is important
 * to note because in an error case we might stay in this code
 * for an extended amount of time. So far, we think this is the
 * best solution, but real-world experience might tell us a
 * different truth ;)
 * rgerhards 2005-10-12
 */
void OnReceive(srAPIObj* pAPI, srSLMGObj* pSLMG)
{
	unsigned char *pszRawMsg;
	int iRetries; /* number of retries connecting to log socket */
	int iSleep;
	int iWriteOffset;
	ssize_t nToWrite;
	ssize_t nWritten;

	srSLMGGetRawMSG(pSLMG, &pszRawMsg);

	/* we need to loop writing the message. At least in
	 * theory, a single write might not send all data to the
	 * syslogd. So we need to continue until everything is written.
	 * Also, we need to check if there are any socket erros, in 
	 * which case we reconect. We will re-try indefinitely, if this
	 * is not acceptable, you need to change the code.
	 * rgerhards 2005-10-12
	 */
	iRetries = 0;
	nToWrite = strlen(pszRawMsg);
	iWriteOffset = 0;
	while(nToWrite != 0) {
		if(LogFile < 0 || !connected)
			openlog();
		if(LogFile < 0 || !connected) {
			/* still not connected, retry */
			if(iRetries > 0) {
				iSleep = (iRetries < 30) ? iRetries : 30;
				/* we sleep a little to prevent a thight loop */
				if(Debug)
					printf("multiple retries connecting to log socket"
					       " - doing sleep(%d)\n", iSleep);
				sleep(iSleep);
			}
			++iRetries;
		} else {
			nWritten = write(LogFile, pszRawMsg, strlen(pszRawMsg));
			if(nWritten < 0) {
				/* error, recover! */
				char errStr[1024];
				printf("error writing to domain socket: %s\r\n", rs_strerror_r(errno, errStr, sizeof(errStr)));
				closelog();
			} else {
				/* prepare for (potential) next write */
				nToWrite -= nWritten;
				iWriteOffset += nWritten;
			}
		}
	}

	if(Debug) {
		static int largest = 0;
		int sz = strlen(pszRawMsg);
		if(sz > largest)
			largest = sz;
		printf("Msg(%d/%d):%s\n\n", largest, sz, pszRawMsg);
	}
}


/* As we are single-threaded in this example, we need
 * one way to shut down the listener running on this
 * single thread. We use SIG_INT to do so - it effectively
 * provides a short-lived second thread ;-)
 */
void doShutdown(int i)
{
	printf("Shutting down rfc3195d. Be patient, this can take up to 30 seconds...\n");
	srAPIShutdownListener(pAPI);
}


/* on the the real program ;) */
int main(int argc, char* argv[])
{
	srRetVal iRet;
	int ch;
	struct sigaction sigAct;

	while ((ch = getopt(argc, argv, "di:np:r:v")) != EOF)
		switch((char)ch) {
		case 'd':		/* debug */
			Debug = 1;
			break;
		case 'i':		/* pid file name */
			PidFile = optarg;
			break;
		case 'n':		/* don't fork */
			NoFork = 1;
			break;
		case 'p':		/* path to regular log socket */
			pPathLogname = optarg;
			break;
		case 'r':		/* listen port */
			listenPort = atoi(optarg);
			if(listenPort < 1 || listenPort > 65535) {
				printf("Error: invalid listen port '%s', using 601 instead\n",
					optarg);
				listenPort = 601;
			}
			break;
		case 'v':
			printf("rfc3195d %s.%s (using liblogging version %d.%d.%d).\n",
			       VERSION, PATCHLEVEL,
			       LIBLOGGING_VERSION_MAJOR, LIBLOGGING_VERSION_MINOR,
			       LIBLOGGING_VERSION_SUBMINOR);
			printf("See http://www.rsyslog.com for more information.\n");
			exit(0);
		case '?':
		default:
			usage();
		}
	if ((argc -= optind))
		usage();

	memset(&sigAct, 0, sizeof(sigAct));
	sigemptyset(&sigAct.sa_mask);
	sigAct.sa_handler = doShutdown;
	sigaction(SIGUSR1, &sigAct, NULL);
	sigaction(SIGTERM, &sigAct, NULL);

	if(!Debug)
	{
		sigAct.sa_handler = SIG_IGN;
		sigaction(SIGINT, &sigAct, NULL);
	}

	if((pAPI = srAPIInitLib()) == NULL)
	{
		printf("Error initializing liblogging - aborting!\n");
		exit(1);
	}

	if((iRet = srAPISetOption(pAPI, srOPTION_BEEP_LISTENPORT, listenPort)) != SR_RET_OK)
	{
		printf("Error %d setting listen port - aborting\n", iRet);
		exit(100);
	}

	if((iRet = srAPISetupListener(pAPI, OnReceive)) != SR_RET_OK)
	{
		printf("Error %d setting up listener - aborting\n", iRet);
		exit(101);
	}

	/* now move the listener to running state. Control will only
	 * return after SIGUSR1.
	 */
	if((iRet = srAPIRunListener(pAPI)) != SR_RET_OK)
	{
		printf("Error %d running the listener - aborting\n", iRet);
		exit(102);
	}

	/** control will reach this point after shutdown */

	srAPIExitLib(pAPI);
	return 0;
}
#endif /* #ifndef FEATURE_RFC3195 - main wrapper */

/*
 * vi:set ai:
 */
